packages/aragon-wrapper/src/index.js

```diff
@@ -228,20 +239,42 @@ export default class Aragon {
     const SET_PERMISSION_EVENT = 'SetPermission'
     const CHANGE_PERMISSION_MANAGER_EVENT = 'ChangePermissionManager'
 
-    const aclObservables = [
-      SET_PERMISSION_EVENT,
-      CHANGE_PERMISSION_MANAGER_EVENT
-    ].map(event =>
-      this.aclProxy.events(event)
-    )
+    const ACL_CACHE_KEY = `acl-${aclAddress}`
+    const REORG_SAFETY_BLOCK_AGE = 100
+
+    // Check if we have cached ACL for this address
+    // Cache object for an ACL: { permissions, blockNumber }
+    const cached = await this.cache.get(ACL_CACHE_KEY, {})
+    const { permissions: cachedPermissions, blockNumber: cachedBlockNumber } = cached
+
+    // When using cache, fetch events from the next block after cache
+    const eventsOptions = cachedPermissions && { fromBlock: (cachedBlockNumber + 1) }
+    const events = this.aclProxy.events(null, eventsOptions)
 
-    // Set up permissions observable
+    const latestBlock$ = this.getLatestBlockNumber()
 
+    // A set for keeping track of the events' blockNumber
+    // -1 is to ensure it passes the first comparisons with event's blockNumber
+    const blockNumbers = new Set([-1])
     // Permissions Object:
-    // app -> role -> { manager, allowedEntities -> [ entities with permission ] }
-    this.permissions = merge(...aclObservables).pipe(
-      // Keep track of all the types of events that have been processed
-      scan(([permissions, eventSet], event) => {
+    // { app -> role -> { manager, allowedEntities -> [ entities with permission ] } }
+    const fetchedPermissions = events.pipe(
+      withLatestFrom(latestBlock$),
+      scan((permissions, [event, latestBlockNumber]) => {
+        const lastBlockProcessed = [...blockNumbers].pop()
+
+        // Cache the permissions when we finished processing a block
+        if (event.blockNumber > lastBlockProcessed) {
+          blockNumbers.add(event.blockNumber)
+
+          // Save fully processed valid (non-zero) blocks that are at least 100 blocks old
+          if (lastBlockProcessed > 0 && lastBlockProcessed < (latestBlockNumber - REORG_SAFETY_BLOCK_AGE)) {
+            // clone the permissions so it can be saved without proxy
+            const permissionsToCache = Object.assign({}, permissions)
+            // cache optimistically without worrying if it succeeded
+            this.cache.set(ACL_CACHE_KEY, { permissions: permissionsToCache, blockNumber: lastBlockProcessed })
+          }
+        }
```

Comment:
With `pairwise()` it does mean we have to be careful to create new `permissions` objects in the scan on each run, otherwise `lastPermissions` and `currentPermissions` will be the same object.

We should also modify `makeAddressMapProxy()` so that it makes a new object with all lowercase keys to create the proxy with.

Reply:
Addressed in https://github.com/aragon/aragon.js/pull/233/commits/7e68be156e2b72550c2578e592d0319fb663877c